vault/token_store.go

```diff
@@ -558,51 +572,96 @@ func (ts *TokenStore) storeCommon(entry *TokenEntry, writeSecondary bool) error
-	te.NumUses -= 1
+	return lock
+}
+
+// UseToken is used to manage restricted use tokens and decrement their
+// available uses. Returns two values: a potentially updated entry or, if the
+// token has been revoked, nil; and whether an error was encountered. The
+// locking here isn't perfect, as other parts of the code may update an entry,
+// but usually none after the entry is already created...so this is pretty
+// good.
+func (ts *TokenStore) UseToken(te *TokenEntry) (*TokenEntry, error) {
+	if te == nil {
+		return nil, fmt.Errorf("invalid token entry provided for use count decrementing")
+	}
 
-	// Revoke the token if there are no remaining uses.
-	// XXX: There is a race condition here with parallel
-	// requests using the same token. This would require
-	// some global coordination to avoid, as we must ensure
-	// no requests using the same restricted token are handled
-	// in parallel.
+	// This case won't be hit with a token with restricted uses because we go
+	// from 1 to -1. So it's a nice optimization to check this without a read
+	// lock.
 	if te.NumUses == 0 {
-		return ts.Revoke(te.ID)
+		return te, nil
+	}
+
+	lock := ts.getTokenLock(te.ID)
+
+	lock.Lock()
+	defer lock.Unlock()
+
+	// Call lookupSalted instead of Lookup to avoid deadlocking since Lookup grabs a read lock
+	te, err := ts.lookupSalted(ts.SaltID(te.ID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to refresh entry: %v", err)
+	}
+	if te == nil {
+		return nil, fmt.Errorf("token entry nil after refreshing to decrement use count; token has likely been used already")
+	}
+
+	// If the token is already being revoked, return nil to indicate that it's
+	// no longer valid
+	if te.NumUses == -1 {
+		return nil, nil
+	}
+
+	// Decrement the count. If this is our last use count, we need to indicate
+	// that this is no longer valid, but revocation is deferred to the end of
+	// the call, so this will make sure that any Lookup that happens doesn't
+	// return an entry.
+	if te.NumUses == 1 {
+		te.NumUses = -1
+	} else {
+		te.NumUses -= 1
 	}
 
 	// Marshal the entry
 	enc, err := json.Marshal(te)
 	if err != nil {
-		return fmt.Errorf("failed to encode entry: %v", err)
+		return nil, fmt.Errorf("failed to encode entry: %v", err)
 	}
 
 	// Write under the primary ID
 	saltedId := ts.SaltID(te.ID)
 	path := lookupPrefix + saltedId
 	le := &logical.StorageEntry{Key: path, Value: enc}
 	if err := ts.view.Put(le); err != nil {
-		return fmt.Errorf("failed to persist entry: %v", err)
+		return nil, fmt.Errorf("failed to persist entry: %v", err)
 	}
-	return nil
+
+	return te, nil
 }
 
-// Lookup is used to find a token given its ID
+// Lookup is used to find a token given its ID. It acquires a read lock, then calls lookupSalted.
 func (ts *TokenStore) Lookup(id string) (*TokenEntry, error) {
 	defer metrics.MeasureSince([]string{"token", "lookup"}, time.Now())
 	if id == "" {
 		return nil, fmt.Errorf("cannot lookup blank token")
 	}
+
+	lock := ts.getTokenLock(id)
```

Comment:
It's a read lock to check for a token with an operation in flight but that is hitting its use limit after. This allows us to not allow additional operations.